ggu/main.c

```diff
@@ -16,11 +16,313 @@
+
+	idx = rss_hash_val / RTE_RETA_GROUP_SIZE;
+	shift = rss_hash_val % RTE_RETA_GROUP_SIZE;
+	queue_id = ggu_conf->rss_conf.reta_conf[idx].reta[shift];
+
+	/* XXX Change mapping queue id to the gk instance id efficiently. */
+	for (i = gk_conf->lcore_start_id; i<= gk_conf->lcore_end_id; i++)
+		if (gk_conf->instances[i - gk_conf->lcore_start_id].
+				rx_queue_front == queue_id)
+			block_id = i - gk_conf->lcore_start_id;
+
+	if (block_id == -1) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"ggu: wrong RSS configuration for GK blocks!\n");
+		return;
+	}
+
+	/*
+	 * Obtain mailbox of that GK block,
+	 * and send the policy decision to the GK block.
+	 */
+	entry.op = GGU_POLICY_ADD;
+	rte_memcpy(&entry.u.ggu.flow, &policy->flow, sizeof(policy->flow));
+	entry.u.ggu.params.state = policy->params.state;
+	entry.u.ggu.params.tx_rate_kb_cycle =
+		rte_be_to_cpu_32(policy->params.tx_rate_kb_cycle);
+	entry.u.ggu.params.grantor_id =
+		rte_be_to_cpu_32(policy->params.grantor_id);
+	entry.u.ggu.params.cap_expire_at =
+		rte_be_to_cpu_64(policy->params.cap_expire_at);
+	entry.u.ggu.params.renewal_step_cycle =
+		rte_be_to_cpu_64(policy->params.renewal_step_cycle);
+
+	send_gk_cmd(block_id, &entry, gk_conf);
+}
+
+static void
+process_single_packet(struct rte_mbuf *pkt, struct ggu_config *ggu_conf)
+{
+	uint8_t j;
+	uint8_t *data;
+	uint8_t *policy_ptr;
+	uint8_t num_policies;
+	uint16_t *proto_ptr;
+	uint16_t ether_type;
+	struct ether_hdr *eth_hdr;
+	struct ipv4_hdr *ip4hdr;
+	struct ipv6_hdr *ip6hdr;
+	struct udp_hdr *udphdr;
+
+	eth_hdr = rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+	ether_type = rte_be_to_cpu_16(eth_hdr->ether_type);
+
+	switch (ether_type) {
+	case ETHER_TYPE_IPv4:
+		ip4hdr = rte_pktmbuf_mtod_offset(pkt, 
+					struct ipv4_hdr *, 
+					sizeof(struct ether_hdr));
+		if (ip4hdr->next_proto_id != IPPROTO_UDP)
+			goto free_packet;
+
+		udphdr = (struct udp_hdr *)&ip4hdr[1];
+		break;
+
+	case ETHER_TYPE_IPv6:
+		ip6hdr = rte_pktmbuf_mtod_offset(pkt, 
+					struct ipv6_hdr *, 
+					sizeof(struct ether_hdr));
+		if (ip6hdr->proto != IPPROTO_UDP)
+			goto free_packet;
+
+		udphdr = (struct udp_hdr *)&ip6hdr[1];
+		break;
+
+	default:
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"ggu: unknown network layer protocol %hu\n",
+			ether_type);
+		goto free_packet;
+		break;
+	}
+
+	if (udphdr->src_port != rte_cpu_to_be_16(GGU_SRC_PORT) ||
+			udphdr->dst_port != rte_cpu_to_be_16(GGU_DST_PORT)) {
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"ggu: unknown udp src port %hu, dst port %hu\n",
+			rte_be_to_cpu_16(udphdr->src_port),
+			rte_be_to_cpu_16(udphdr->dst_port));
+		goto free_packet;
+	}
+
+	data = (uint8_t *)&udphdr[1];
+	num_policies = data[0];
+	data = data + 1;
+	proto_ptr = (uint16_t *)data;
+	policy_ptr = (uint8_t *)((uint16_t *)data + num_policies);
+
+	/* Loop over each policy decision on the packet. */
+	for (j = 0; j < num_policies; j++) {
+		int ret;
+		struct ggu_policy policy;
+
+		ret = policy_parse_next(proto_ptr, policy_ptr, &policy);
```

Comment:
Very good catch, I have moved the update outside of the ``policy_parse_next()`` function.